File: Tree/222. Count Complete Tree Nodes/count.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def leftheight(self, root:Optional[TreeNode])-> int:
        if(root==None):
            return 0
        return self.leftheight(root.left)+1

    def rightheight(self, root:Optional[TreeNode])->int:
        if(root==None):
            return 0
        return self.rightheight(root.right)+1

    def countNodes(self, root: Optional[TreeNode]) -> int:
        if(root==None):
            return 0
        left = self.leftheight(root)
        right = self.rightheight(root)
        if(left==right):
            return 2**left-1
        return 1+self.countNodes(root.left)+self.countNodes(root.right)
        

File: Tree/222. Count Complete Tree Nodes/count.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution 
{
public:
    int leftmostheight(TreeNode* root)
    {
        if(root==NULL)
        {
            return 0;
        }
        int l = leftmostheight(root->left);
        return l+1;
    }
    int rightmostheight(TreeNode* root)
    {
        if(root==NULL)
        {
            return 0;
        }
        int r = rightmostheight(root->right);
        return r+1;
    }
    int countNodes(TreeNode* root) 
    {
        
        if(root==nullptr)
        {
            return 0;
        }
        int left=leftmostheight(root);
        int right=rightmostheight(root);
        if(left==right)
        {
           return pow(2,left)-1;
        }
        return 1+countNodes(root->left)+countNodes(root->right);
    }
};